beacon-chain/core/altair/epoch_precompute.go

```diff
@@ -163,3 +165,120 @@ func ProcessEpochParticipation(
+		return state, nil
+	}
+
+	numOfVals := state.NumValidators()
+	// Guard against an out-of-bounds using validator balance precompute.
+	if len(vals) != numOfVals || len(vals) != state.BalancesLength() {
+		return state, errors.New("validator registries not the same length as state's validator registries")
+	}
+
+	attsRewards, attsPenalties, err := AttestationsDelta(state, bal, vals)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get attestation delta")
+	}
+
+	balances := state.Balances()
+	for i := 0; i < numOfVals; i++ {
+		vals[i].BeforeEpochTransitionBalance = balances[i]
+
+		// Compute the post balance of the validator after accounting for the
+		// attester and proposer rewards and penalties.
+		balances[i] = helpers.IncreaseBalanceWithVal(balances[i], attsRewards[i])
+		balances[i] = helpers.DecreaseBalanceWithVal(balances[i], attsPenalties[i])
+
+		vals[i].AfterEpochTransitionBalance = balances[i]
+	}
+
+	if err := state.SetBalances(balances); err != nil {
+		return nil, errors.Wrap(err, "could not set validator balances")
+	}
+
+	return state, nil
+}
+
+// AttestationsDelta computes and returns the rewards and penalties differences for individual validators based on the
+// voting records.
+func AttestationsDelta(state state.BeaconStateAltair, bal *precompute.Balance, vals []*precompute.Validator) (rewards, penalties []uint64, err error) {
+	numOfVals := state.NumValidators()
+	rewards = make([]uint64, numOfVals)
+	penalties = make([]uint64, numOfVals)
+	prevEpoch := helpers.PrevEpoch(state)
+	finalizedEpoch := state.FinalizedCheckpointEpoch()
+
+	for i, v := range vals {
+		rewards[i], penalties[i] = attestationDelta(bal, v, prevEpoch, finalizedEpoch)
+	}
+
+	return rewards, penalties, nil
+}
+
+func attestationDelta(bal *precompute.Balance, v *precompute.Validator, prevEpoch, finalizedEpoch types.Epoch) (reward, penalty uint64) {
+	eligible := v.IsActivePrevEpoch || (v.IsSlashed && !v.IsWithdrawableCurrentEpoch)
+	// Per spec `ActiveCurrentEpoch` can't be 0 to process attestation delta.
+	if !eligible || bal.ActiveCurrentEpoch == 0 {
+		return 0, 0
+	}
+
+	cfg := params.BeaconConfig()
+	increment := cfg.EffectiveBalanceIncrement
+	factor := cfg.BaseRewardFactor
+	effectiveBalance := v.CurrentEpochEffectiveBalance
+	baseReward := (effectiveBalance / increment) * (increment * factor / mathutil.IntegerSquareRoot(bal.ActiveCurrentEpoch))
+	activeIncrement := bal.ActiveCurrentEpoch / increment
+
+	weightDenominator := cfg.WeightDenominator
+	srcWeight := cfg.TimelySourceWeight
+	tgtWeight := cfg.TimelyTargetWeight
+	headWeight := cfg.TimelyHeadWeight
+	reward, penalty = uint64(0), uint64(0)
+	// Process source reward / penalty
+	inactivityLeak := helpers.IsInInactivityLeak(prevEpoch, finalizedEpoch)
+	if v.IsPrevEpochAttester && !v.IsSlashed {
+		if !inactivityLeak {
+			n := baseReward * srcWeight * (bal.PrevEpochAttested / increment)
+			reward += n / (activeIncrement * weightDenominator)
```

Comment:
Thought about refactoring this but I'm not sure if it's better given the amount of parameters

```go
reward += getReward(baseReward, srcWeight, bal.PrevEpochAttested, activeIncrement)

func getReward(baseReward, weight, attestedAmount, activeIncrement uint64) { }
```